java/ql/src/experimental/Security/CWE/CWE-094/SpringViewManipulation.ql

```diff
@@ -13,7 +13,15 @@ import java
 import SpringViewManipulationLib
 import DataFlow::PathGraph
 
-from DataFlow::PathNode source, DataFlow::PathNode sink, SpringViewManipulationConfig conf
-where conf.hasFlowPath(source, sink)
+from
+  DataFlow::PathNode source, DataFlow::PathNode sink, SpringViewManipulationConfig conf,
+  StringFlowConfig s
+where
+  thymeleafIsUsed() and
+  conf.hasFlowPath(source, sink) and
+  not s.hasFlow(_, sink.getNode()) and
+  not exists(AddExpr e | e.getLeftOperand().(StringLiteral).getRepresentedString() = "redirect:" |
+    e = sink.getNode().asExpr()
```

Comment:
is this necessary given the StringFlowConfig?

Reply:
yes, codeql won't recognize the local flow otherwise. I tried testing with the dataflow config alone. QL won't recognise the flow from in the case of statements like  
```java
return "redirect:" + request.getHeader("Referer");
```

Reply:
thanks for the info, it make sense since both the source and the sink belong to the same expression. ccing @aschackmull to double check

Reply:
Well, the `StringFlowConfig` likely won't do much, as it only does pure data flow without any taint steps, so it'll only report flow if the literal string "redirect:" (or "ajaxredirect:") reaches the sink unmodified. Adding a suitable string concatenation step as an additional step would address this.
However, checking sanitization through a separate flow config like this likely isn't what you want, since if a sink is reached both by `"redirect:" + something` and remotely controlled data, without the former path being a subpath of the latter, then you still have a vulnerability, but this sort of pattern will remove those results leading to false negatives.
Instead there's a much simpler and better solution: Get rid of `StringFlowConfig` and insert the concatenated-with-"redirect:" check as a sanitizer directly in the `SpringViewManipulationConfig`, i.e. add something like
```
override predicate isSanitizer(DataFlow::Node node) {
  exists(AddExpr e |
    node.asExpr() = e and
    e.getLeftOperand().(StringLiteral).getRepresentedString() = ["redirect:", "ajaxredirect:"]
  )
}
```
This will remove all such `AddExpr` nodes from the data-flow graph in the calculation of flow in `SpringViewManipulationConfig`.

Reply:
Thanks @aschackmull! @porcupineyhairs do not forget to account for `Appendable.append` (StringBuffer, StringBuilder, StringWriter, ...) and `String.format`

Reply:
@aschackmull I considered adding the `isSanitizer` check earlier. String concatenation steps such as those suggested by @pwntester  would make writing a good sanitizer almost impossible. Hence I chose the flow config. I could change it to a taintflow to fix the issue you pointed earlier. WDYT?

Reply:
I don't think it sounds that difficult to incorporate those additional kinds of string concatenation in a sanitizer. `String.format` should be very similar to an `AddExpr`, and for things like `StringBuilder` you can likely just mark any local flow from a `StringBuilder` that gets a string like `"redirect"` as being sanitizers, i.e. all occurrences of `x` in the following:
```
x.append("redirect:");
x.append(tainted());
return x.toString();
```
As you can see, this will block flow from `tainted()` reaching the `return` when the final `x` is marked as a sanitizer.

Reply:
@aschackmull @pwntester I have added the sanitisers and rebased to the latest main.